### ISSUE TITLE
Tokens: Revamp locking & reserving

### ISSUE THREAD

**Issue by gavofyork** [paritytech/polkadot-sdk#236]
We want to revisit the definitions of free, locked and reserved, especially with regards voting, auctions/crowdloans and staking. Staking and bidding should be mutually exclusive but either should be possible at the same time as (conviction) voting. No funds used for any of the three should be transferable. We should be aware that funds for staking may be slashed.

The reorganisation will be multifaceted. Terminologically, we will move away from "reserve" and instead use the word "hold"; "locking" will become "freezing". When placing a hold on funds, the caller must include a `HoldReason` value which is essentially equivalent to a name when using named reserves except that it is typesafe. Slashing will not be possible except on funds on hold. Freezing will be possible on the total balance, including any funds on hold. Finally, we will deprecate the old `*Currency` traits and move all core pallets over to use `fungible::*` traits.

## 1. Typesafe reserve/hold identifier for the reasons for holding (reserving).

We should have proper reserve identifiers rather than just strings/blobs. Within a pallet which uses holds or slashes, you must do something like:

```
/// Reason why this pallet would want to place funds on hold. This will often only have one
/// entry since most pallets only have one call-point (and thus reason) for placing funds
/// on hold (in reserve).
#[pallet::hold_reason]
#[derive(Debug, TypeInfo, Encode, Decode /* etc */)]
pub enum HoldReason {
    /// Funds are on hold for staking.
    Staking,
}
```

They'd be aggregated into a top-level type (like `Error`, `Event`, `Origin`, &c.), and that would be exposed in metadata. You'd then be required to provide a value of this type when calling `hold`. This would make it basically impossible for one pallet to `slash` or `unreserve` another pallet's funds.

### Lock IDs, also...

In addition to an identifier for holds, there should be another one in parallel for identifying locks (right now it's just a blob).

## 2. Change lock/reserve/hold mechanics

- Make locking apply to the total balance, not just the free balance, so funds on hold can also be part of the lock.
- Ensure that the ED applies to the free balance, not any balance on hold.
- Remove `hold`/`reserve` and rename `reserve_named`/`hold_named` to `reserve`/`hold`; this makes a reason/name/ID be required for any funds placed on hold.
- Other functions which operate on funds on hold must also accept a reason, including for repatriation (which, like slashing, ignores locks) and transfer (which respects locks).
- Remove the `slash` and `slash_reserved` functions; free and locked funds are then non-slashable and all slashes must have a "reason".
- Rename `slash_reserved_named` to `slash`; again, this cements the requirement that only segregated, pre-reserved funds are slashable. (If you just want to slash free funds which you know are there, then you can still reserve them and slash them.)

Reserving funds basically becomes a "pre-slash lock" with a hard guarantee that until you unreserve them, they're definitely slashable.

## 3. Update usages

- Switch the Staking system to hold funds, rather than merely locking them.
- Avoid having the PLA bidding system from reserving itself, but have it trust the caller that the funds have been reserved properly. The Crowdloan pallet would reserve them in place in the user account. There would need to be extra logic for bids which don't come from the Crownloan pallet, which would need to do the reserve/unreserve operation.

All non-deprecated pallets which use locks should have some unpermissioned dispatchable which allows the bookkeeping to be updated in case an account's reserved funds received a slash and there is no longer the expected funds locked. For now this would basically just be `conviction-voting` pallet.

## 4. Move to `fungible::*`

Deprecate `Currency`/`LockableCurrency`/`ReservableCurrency` traits to the `fungible::*` traits. Doing all of this will need a migration to move staked funds from a lock to a named reserve, but it should be possible. Migrations will also need to happen for any pallets which use `reserve` without a name. Migrations will also be needed for the PLA system (for the new system of holds) and the staking system (to move from locks to holds).

## 5. Alter NIS

NIS doesn't really fit into this model. We should change it to allow the receipt-holder to retain the funds under a hold. If the receipt gets transferred then the underlying funds will also need to be transferred (not repatriated - we want to respect any locks such as for voting). If the funds transfer fails, then the receipt transfer should also fail.
